pallets/api/src/messaging/mod.rs

```diff
@@ -234,151 +249,174 @@ pub mod pallet {
-			);
 			T::Deposit::hold(&HoldReason::Messaging.into(), &origin, deposit)?;
 
 			// Process message by dispatching request via ISMP.
-			ensure!(!Messages::<T>::contains_key(&origin, &id), Error::<T>::MessageExists);
-			let commitment = T::IsmpDispatcher::default()
-				.dispatch_request(message.into(), FeeMetadata { payer: origin.clone(), fee })
-				.map_err(|_| Error::<T>::DispatchFailed)?;
-
-			// Store commitment for lookup on response, message for querying, response/timeout
-			// handling.
-			IsmpRequests::<T>::insert(&commitment, (&origin, id));
-			Messages::<T>::insert(&origin, id, Message::Ismp { commitment, callback, deposit });
-			Pallet::<T>::deposit_event(Event::<T>::IsmpGetDispatched {
-				origin,
-				id,
-				commitment,
-				callback,
-			});
-			Ok(())
+			let maybe_commitment = T::IsmpDispatcher::default()
+				.dispatch_request(message.into(), FeeMetadata { payer: origin.clone(), fee });
+
+			match maybe_commitment {
+				Ok(commitment) => {
+					// Store commitment for lookup on response, message for querying,
+					// response/timeout handling.
+					IsmpRequests::<T>::insert(&commitment, (&origin, id));
+					Messages::<T>::insert(
+						&origin,
+						id,
+						Message::Ismp {
+							commitment,
+							callback: callback.clone(),
+							deposit,
+							status: MessageStatus::Ok,
+						},
+					);
+					Pallet::<T>::deposit_event(Event::<T>::IsmpGetDispatched {
+						origin,
+						id,
+						commitment,
+						callback,
+					});
+					Ok(())
+				},
+				Err(_) => {
+					// Allow a caller to poll for the status still and retreive the message deposit.
+					Messages::<T>::insert(
+						&origin,
+						id,
+						Message::Ismp {
+							commitment: Default::default(),
+							callback,
+							deposit,
+							status: MessageStatus::Err(Error::<T>::IsmpDispatchFailed.into()),
+						},
+					);
+					Err(Error::<T>::IsmpDispatchFailed.into())
+				},
+			}
 		}
 
 		// TODO: does ismp allow querying to ensure that specified para id is supported?
 		#[pallet::call_index(2)]
-		#[pallet::weight(Weight::zero() // todo: benchmarking after consolidating storage
-			// Add any additional gas limit specified for callback execution
-			.saturating_add(callback.map(|cb| {
-				T::CallbackExecutor::execution_weight().saturating_add(cb.weight)
-			}).unwrap_or_default())
-		)]
+		#[pallet::weight(Weight::zero())]
 		pub fn ismp_post(
 			origin: OriginFor<T>,
 			id: MessageId,
 			message: ismp::Post<T>,
 			fee: BalanceOf<T>,
-			callback: Option<Callback>,
+			callback: Option<Callback<T::AccountId>>,
 		) -> DispatchResult {
 			let origin = ensure_signed(origin)?;
+			ensure!(!Messages::<T>::contains_key(&origin, &id), Error::<T>::MessageExists);
+
+			let deposit = calculate_protocol_deposit::<T, T::OnChainByteFee>(
+				ProtocolStorageDeposit::IsmpRequests,
+			) + calculate_message_deposit::<T, T::OnChainByteFee>() +
+				calculate_deposit_of::<T, T::OffChainByteFee, ismp::Post<T>>();
```

Comment:
Should we use safe arithmetic here (saturating add or checked add)?

Reply:
Yes absolutely and test, second iteration will cover this